### ISSUE TITLE
allow defining multiple (row + column) permission tuples for a given role + table + action

### ISSUE THREAD

**Issue by rikinsk** [hasura/graphql-engine#3442]
To restrict access to sensitive fields, the solution proposed in https://docs.hasura.io/1.0/graphql/manual/queries/control-access.html  requires one to modify the db schema (creating a view and setting up new relationships/permissions) 

Instead if it were possible to define multilple row + column permission rules this could be avoided. 

e.g. For table `user(id, name, email, phone)`, `select` permissions for role `user` can be

- Rule 1:
  - Row permission: `{}` i.e. without any checks
  - Col permission: `name`

- Rule 2:
  - Row permission: `{id: {_eq: 'x-hasura-user-id'}}` i.e. if user is the owner
  - Col permission: `name, email, phone`

**Comment by yasinarik** [comment]
This is the exact need I had 2 days ago. I would solve this by moving sensitive & private data to a new table at the moment. However, if multiple rules for the same role in a table can be implemented, it will become easier to control & develop.

**Current Solution** 
Also, even though I did a research about it, I couldn't reach the above documentation page which suggests creating a view by using: 
```
Data -> SQL


CREATE VIEW user_public AS
  SELECT id, name, email
    FROM user_profile;
```

where the original fields of a user profile are:
```
user_profile (id, name, email, phone, address)
```

**Comment by tirumaraiselvan** [comment]
Related to #877

**Comment by bertyhell** [comment]
My usecase is similar to this, but a bit more complex.

eg:
* Rule 1:
** Row permission: {} i.e. without any checks
** Col permission: name

Rule 2:
** Row permission: {id: {_eq: 'x-hasura-user-id'}} i.e. if user is the owner
** Col permission: email

Rule 3:
** Row permission: {id: {_eq: 'x-hasura-user-id'}, {permissions: {name: {_eq: "CAN_VIEW_PHONE"}}}} i.e. if user has certain permissions linked to its user account
** Col permission: phone

Ideally, all checks would be run and the union of all columns would be used as the final result.

eg: if the user id matches but the user does not have the CAN_VIEW_PHONE permission, then he would be allowed to access the columns: ["name", "email"]

**Comment by webdeb** [comment]
Guys let me get some attention to the small rfc I've created here https://github.com/hasura/graphql-engine/pull/5783 where I am touching exactly this problem, with a slightly other mindset. Instead of multiple rules, the idea is to define virtual roles
role + check + columns.

A virtual role is a role which a user has for this particular table, like I am the _author_ of this post. Or I am collaborator.

So instead of extending the single role capabilities with multiple rules inside a role, we ca have many virtual roles for the same table: a user can have different relationships to a table, like _author_, _owner_, _moderator_ etc.

To control which vrole should be used, we can use graphql query itself and the client should tell the engine _as_ who he is gonna to interact with the table. like `user(_as: me) { my_private_info }`

If you curious, please check out the rfc, maybe I could contribute to this and learn some haskell btw :)

**Comment by webdeb** [comment]
The lack of this feature does not let me sleep well, so here another idea:

add a suffix to the role to define which rule you want to use:

![Bildschirmfoto 2020-11-04 um 18 00 04](https://user-images.githubusercontent.com/14992140/98142579-4384c800-1ec8-11eb-9c98-a05cd7b203ab.png)

I am not sure how to best implement this. I guess the simplest would be just with the "x-hasura-role" Header, but you might want to use different rules on different tables so with the `_as` option it would be theoretically possible:

```
user(_as: "me") {
  balance

  projects(_as: "contributor") {
    name
  }
}
```

**Comment by webdeb** [comment]
would love to get some attention from you guys here: @rikinsk @tirumaraiselvan 

because right now you have to create a separate view to support multiple permissions on the same table, instead of multiple rules on the same table

It's very cool that hasura supports views and functions, but using them to slice permissions is a workaround.

**Comment by hugbubby** [comment]
> To restrict access to sensitive fields, the solution proposed in https://docs.hasura.io/1.0/graphql/manual/queries/control-access.html requires one to modify the db schema (creating a view and setting up new relationships/permissions)
> 
> Instead if it were possible to define multilple row + column permission rules this could be avoided.
> 
> e.g. For table `user(id, name, email, phone)`, `select` permissions for role `user` can be
> 
>     * Rule 1:
>       
>       * Row permission: `{}` i.e. without any checks
>       * Col permission: `name`
> 
>     * Rule 2:
>       
>       * Row permission: `{id: {_eq: 'x-hasura-user-id'}}` i.e. if user is the owner
>       * Col permission: `name, email, phone`

This is a very big limitation on how permissions work rn, and it's a little scary there hasn't been much activity here in over a year

**Comment by tirumaraiselvan** [comment]
Folks interested in this issue, pls do attend our community call today (will also be recorded) where we showcase the brand new "inherited roles" feature, which will let you combine 2 roles into one. And as a consequence solve the problem of multiple rules as well!

**Comment by webdeb** [comment]
@tirumaraiselvan thank you for updating. Would be interesting if we could create "virtual" roles, which are not part of the JWT as global roles-set but rather a role which exists only for a particular table/view. 

For example we can have application global roles like:  user & moderator (& anonymous).

However, a "user" can have local roles for a table, like the "manager" of a project. Or "topic editor" on posts.
But this roles are not included in the JWT, because they are dynamic, based on data and not on the users authentication profile. they are all inherited from the base "user" role with additional check queries and column permissions read & write.

I'll try to watch the community call today, just wanted to use the opportunity while you are watching this issue.

**Comment by webdeb** [comment]
@tirumaraiselvan I don't think that inherited roles solves this issue. You cannot define **multiple permission tuples for a given role + table + action**

**Comment by tirumaraiselvan** [comment]
> @tirumaraiselvan I don't think that inherited roles solves this issue. You cannot define **multiple permission tuples for a given role + table + action**

Ah yes, it doesn't directly. But you can solve it indirectly - for each permission tuple, create one role and then create an inherited role which inherits all those roles. Sorry, should have been clearer.

**Comment by webdeb** [comment]
@tirumaraiselvan The problem with you proposal is that you have to create this roles globally. And permission rules are actually located to some table, they define the relationship of the user-id and a table row.

Anyway, I see that with the underlying technique of conditional column selection and multiple roles it might be possible to have inverted inherited roles per table/resource.

Just let us have permission rules as combination of role + : + rule. For example `user:manager`, `user:employee` on the `timestamping` table. This would automatically generate the roles `user_manager` & `user_employee` & and the inherited role `user` which is already part. And to make it backwards compatible generate from `user` the role `user_default` and the inherited role `user`.

However, right now if I create a role on a table, this role is created globally. To have multiple rules per role & table  "user:owner" should be only local to the actual table. Otherwise it just wont scale.

**Comment by Mythicaeda** [comment]
Adjacent issue to the `select` permissions discussed so far: needing multiple tuples for `update` permissions.

Scenario: I've got a Document table that has a `name`, an `owner`, and a collection of `editor`s. A `user` can be an `owner` on one document and an `editor` on another document (similar to Google Docs).

I need to be able to represent the following sets of permissions:

- Rule 1: "Document Owner can change the Name and Owner of the Document":
    -    Row Permissions: `{owner: {_eq: { 'x-hasura-user-id'}}` AKA if user owns the document
    -    Column Permissions: `[name, owner]` 
- Rule 2: "Document Editor can only change the Name of the Document": 
    -    Row Permissions: `{editors: {_editor: {_eq: { 'x-hasura-user-id'}}}` AKA if user is an editor
    -    Column Permissions: `[name]` 

Attempting the solution @tirumaraiselvan proposed (creating an `owner` and `editor` role that follow Rule 1 and Rule 2 respectively, then creating a `user` role that inherits from `owner` and `editor`) triggers an `inherited role permission inconsistency` Metadata error, which suggests that this is **not** currently solvable via Inherited Roles.
